pkg/pillar/cmd/zedrouter/flowstats.go

```diff
@@ -0,0 +1,738 @@
+		timeout           = 10 * time.Second // collect enough packets in 10sec before processing
+		handle      *pcap.Handle
+		filter      = "udp and port 53"
+	)
+	if bnNum >= maxBridgeNumber {
+		log.Errorf("Can not snoop on brige number %d", bnNum)
+		return
+	}
+	log.Infof("(FlowStats) DNS Monitor on %s(bridge-num %d)", bn, bnNum)
+
+	handle, err = pcap.OpenLive(bn, snapshotLen, promiscuous, timeout)
+	if err != nil {
+		log.Errorf("Can not snoop on bridge %s", bn)
+		return
+	}
+	defer handle.Close()
+
+	err = handle.SetBPFFilter(filter)
+	if err != nil {
+		log.Errorf("Can not install DNS filter on %s", bn)
+		return
+	}
+
+	dnssys[bnNum].Done = make(chan bool)
+	packetSource := gopacket.NewPacketSource(handle, handle.LinkType())
+	dnsIn := packetSource.Packets()
+	for {
+		var packet gopacket.Packet
+		select {
+		case <-dnssys[bnNum].Done:
+			log.Infof("(FlowStats) DNS Monitor exit on %s(bridge-num %d)", bn, bnNum)
+			defer close(dnssys[bnNum].Done)
+			return
+		case packet = <-dnsIn:
+			dnssys[bnNum].Lock()
+			checkDNSPacketInfo(bnNum, packet)
+			dnssys[bnNum].Unlock()
+		}
+	}
+}
+
+// DNSStopMonitor : Stop DNS Query monitoring
+func DNSStopMonitor(bnNum int) {
+	log.Infof("(FlowStats) Stop DNS Monitor on bridge-num %d", bnNum)
+	dnssys[bnNum].Done <- true
+	dnssys[bnNum].Lock()
+	dnsDataRemove(bnNum)
+	dnssys[bnNum].Unlock()
+}
+
+func checkDNSPacketInfo(bnNum int, packet gopacket.Packet) {
+	var DstIP net.IP
+	var dnsentry dnsEntry
+	ipLayer := packet.Layer(layers.LayerTypeIPv4)
+	if ipLayer != nil {
+		ip, _ := ipLayer.(*layers.IPv4)
+		DstIP = ip.DstIP
+		dnsentry.isIPv4 = true
+	} else {
+		ipv6Layer := packet.Layer(layers.LayerTypeIPv6)
+		if ipv6Layer != nil {
+			ipv6, _ := ipv6Layer.(*layers.IPv6)
+			DstIP = ipv6.DstIP
+		}
+	}
+
+	dnsLayer := packet.Layer(layers.LayerTypeDNS)
+	if dnsLayer != nil {
+		dns, _ := dnsLayer.(*layers.DNS)
+		if dns.ANCount > 0 {
+			dnsentry.AppIP = DstIP
+			for _, dnsQ := range dns.Questions {
+				var checkProto, haveAN bool
+				dnsentry.DomainName = string(dnsQ.Name)
+				dnsentry.TimeStamp = time.Now()
+				dnsentry.ANCount = dns.ANCount
+				for _, dnsA := range dns.Answers {
+					if dnsA.Type != layers.DNSTypeA && dnsA.Type != layers.DNSTypeAAAA { // only for A or AAAA
+						continue
+					}
+					if dnsA.IP.String() != "" {
+						if checkProto == false {
+							dnsentry.isIPv4 = dnsAddrIsIPv4(dnsA.IP)
+							checkProto = true
+						}
+						dnsentry.Answers = append(dnsentry.Answers, dnsA.IP)
+						haveAN = true
+					}
+				}
+				if haveAN {
+					dnssys[bnNum].Snoop = append(dnssys[bnNum].Snoop, dnsentry)
+					log.Infof("!!--FlowStats: DNS collected for %s", string(dnsQ.Name))
+					break
+				}
+			}
+		}
+	}
+}
+
+func dnsDataRemove(bnNum int) {
+	if len(dnssys[bnNum].Snoop) > 0 {
+		dnssys[bnNum].Snoop = nil
+	}
+}
+
+func dnsAddrIsIPv4(addr net.IP) bool {
+	if addr != nil {
+		if strings.Contains(addr.String(), ":") {
```

Comment:
Please use net.To4() != nil to check if it is IPv4.

Reply:
will change.